## Building a Model for Classifying Animal Type and Assessing Its Performance

In [1]:
#2
import pandas as pd

In [2]:
#3
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter04/Dataset/openml_phpZNNasq.csv'

In [3]:
#4/5
df = pd.read_csv(file_url)
df.head()

animal   hair  feathers   eggs   milk  airborne  aquatic  predator  \
0  aardvark   True     False  False   True     False    False      True   
1  antelope   True     False  False   True     False    False     False   
2      bass  False     False   True  False     False     True      True   
3      bear   True     False  False   True     False    False      True   
4      boar   True     False  False   True     False    False      True   

   toothed  backbone  breathes  venomous   fins  legs   tail  domestic  \
0     True      True      True     False  False     4  False     False   
1     True      True      True     False  False     4   True     False   
2     True      True     False     False   True     0   True     False   
3     True      True      True     False  False     4  False     False   
4     True      True      True     False  False     4   True     False   

   catsize    type  
0     True  mammal  
1     True  mammal  
2    False    fish  
3     True  mammal  
4     True  mammal

In [4]:
#6 
# We will be using type as our target variable
# remove animal column using the .drop()
df.drop(columns='animal', inplace=True)

In [5]:
#7/8
# extract type column using .pop()
y = df.pop('type')
df.head()

hair  feathers   eggs   milk  airborne  aquatic  predator  toothed  \
0   True     False  False   True     False    False      True     True   
1   True     False  False   True     False    False     False     True   
2  False     False   True  False     False     True      True     True   
3   True     False  False   True     False    False      True     True   
4   True     False  False   True     False    False      True     True   

   backbone  breathes  venomous   fins  legs   tail  domestic  catsize  
0      True      True     False  False     4  False     False     True  
1      True      True     False  False     4   True     False     True  
2      True     False     False   True     0   True     False    False  
3      True      True     False  False     4  False     False     True  
4      True      True     False  False     4   True     False     True

In [6]:
#9 import train_test_split function
from sklearn.model_selection import train_test_split

In [7]:
#10 split the dataset into training and testing sets
X_train,X_test,y_train,y_test = train_test_split(df, y, test_size=0.4, random_state=188)

In [8]:
#11 import RandomForestClassifier from sklearn.ensemble
from sklearn.ensemble import RandomForestClassifier
#12 instanstiate the RandomForestClassifier object with random_state 42; set the n-estimators value to an initial default 
    # value of 10
rf_model = RandomForestClassifier(random_state=42, n_estimators=10)

In [9]:
#13 fit RandomForestClassifier with the training set
rf_model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=10, random_state=42)

In [10]:
#14 predict the outcome of the training set
train_preds = rf_model.predict(X_train)
train_preds

array(['mammal', 'mammal', 'mammal', 'fish', 'mammal', 'insect', 'fish',
       'bird', 'mammal', 'mammal', 'fish', 'bird', 'reptile', 'bird',
       'fish', 'mammal', 'mammal', 'bird', 'bird', 'mammal', 'bird',
       'bird', 'mammal', 'invertebrate', 'reptile', 'invertebrate',
       'fish', 'bird', 'mammal', 'mammal', 'amphibian', 'mammal',
       'invertebrate', 'mammal', 'mammal', 'insect', 'mammal', 'fish',
       'invertebrate', 'mammal', 'invertebrate', 'invertebrate', 'insect',
       'amphibian', 'mammal', 'reptile', 'amphibian', 'invertebrate',
       'mammal', 'fish', 'bird', 'mammal', 'mammal', 'bird', 'mammal',
       'mammal', 'fish', 'mammal', 'bird', 'fish'], dtype=object)

In [11]:
#15 import the accuracy score from sklearn.metrics
from sklearn.metrics import accuracy_score

#16 calculate accuracy score on the training set
train_acc = accuracy_score(y_train, train_preds)
print(train_acc)

1.0


In [12]:
#17 check the performance on testing set
test_preds = rf_model.predict(X_test)

In [13]:
#18 calculate the accuracy score on testing set
test_acc = accuracy_score(y_test, test_preds)
print(test_acc)

0.8780487804878049


In this exercise, we trained a RandomForest to predict the type of animals based on their key attributes. Our model achieved a perfect accuracy score of 1 on the training set but only 0.88 on the testing set. This means our model is overfitting and is not general enough. The ideal situation would be for the model to achieve a very similar, high-accuracy score on both the training and testing sets.